### ISSUE TITLE
Intermittently get NoSuchConnectionException while using a jdbc client on a micro instance

### ISSUE THREAD

**Issue by pcallahan-r7** [apache/druid#11855]
Please provide a detailed title (e.g. "Broker crashes when using TopN query with Bound filter" instead of just "Broker crashes").

### Affected Version

0.22.0

### Description

Running a druid micro instance.   I'm using the avatica jdbc driver (1.18.0) on the client side to make SQL queries to druid.  

About 50% of the calls result in an "UncategorizedSQLException" on the client.  On the druid side, there is this exception:

```
[org.apache.calcite.avatica.NoSuchConnectionException
	at org.apache.druid.sql.avatica.DruidMeta.getDruidConnection(DruidMeta.java:590)
	at org.apache.druid.sql.avatica.DruidMeta.createStatement(DruidMeta.java:139)
	at org.apache.druid.sql.avatica.DruidMeta.prepare(DruidMeta.java:150)
	at org.apache.calcite.avatica.remote.LocalService.apply(LocalService.java:195)
	at org.apache.calcite.avatica.remote.Service$PrepareRequest.accept(Service.java:1215)
	at org.apache.calcite.avatica.remote.Service$PrepareRequest.accept(Service.java:1186)
	at org.apache.calcite.avatica.remote.AbstractHandler.apply(AbstractHandler.java:94)
	at org.apache.calcite.avatica.remote.JsonHandler.apply(JsonHandler.java:52)
	at org.apache.calcite.avatica.server.AvaticaJsonHandler.handle(AvaticaJsonHandler.java:133)
	at org.apache.druid.sql.avatica.DruidAvaticaJsonHandler.handle(DruidAvaticaJsonHandler.java:59)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:59)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:179)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:383)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:882)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036)
	at java.lang.Thread.run(Thread.java:748)
]
```

The SQL queries are simple and they work about 50% of the time.    Also using Spring Hikari connection pool, if that matters.

**Comment by winterbe** [comment]
The cause of this exception (at least in my case) was connection timeouts on the Druid side. If you execute a query, then idle for a few minutes and execute the same query again, the exception is thrown. 

The solution when using the Hikari Connection Pool is quite simple. You have to set `keepaliveTime` for the connection pool, e.g. `keepaliveTime=30000`.

See also Spring Boot property: [spring.datasource.hikari.keepalive-time](https://docs.spring.io/spring-boot/docs/current/reference/html/application-properties.html#application-properties.data.spring.datasource.hikari)

**Comment by lthoulon-locala** [comment]
Hello. I'm having the same issue and even setting the `keepalive-time` it's failing after 4 minutes of idle time. I guess this is linked to the `unusedConnectionTimeout`.

I do however see the keepalive working on my second connection.
```
2023-10-18T17:24:16.753+02:00 DEBUG 18188 --- [l-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Pool stats (total=5, active=0, idle=5, waiting=0)
2023-10-18T17:24:16.754+02:00 DEBUG 18188 --- [l-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Fill pool skipped, pool has sufficient level or currently being filled (queueDepth=0).
2023-10-18T17:24:20.002+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - Pool stats (total=5, active=0, idle=5, waiting=0)
2023-10-18T17:24:20.002+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - Fill pool skipped, pool has sufficient level or currently being filled (queueDepth=0).
2023-10-18T17:24:26.452+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@52f0bb54 is alive
2023-10-18T17:24:29.238+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@36919a6d is alive
2023-10-18T17:24:30.369+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@652279f0 is alive
2023-10-18T17:24:31.113+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@190ba0b6 is alive
2023-10-18T17:24:32.424+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@b89cbf9 is alive
2023-10-18T17:24:46.825+02:00 DEBUG 18188 --- [l-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Pool stats (total=5, active=0, idle=5, waiting=0)
2023-10-18T17:24:46.826+02:00 DEBUG 18188 --- [l-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Fill pool skipped, pool has sufficient level or currently being filled (queueDepth=0).
2023-10-18T17:24:50.075+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - Pool stats (total=5, active=0, idle=5, waiting=0)
2023-10-18T17:24:50.075+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - Fill pool skipped, pool has sufficient level or currently being filled (queueDepth=0).
2023-10-18T17:24:54.356+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@52f0bb54 is alive
2023-10-18T17:24:57.357+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@36919a6d is alive
2023-10-18T17:24:58.540+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@652279f0 is alive
2023-10-18T17:24:59.370+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@190ba0b6 is alive
2023-10-18T17:25:00.905+02:00 DEBUG 18188 --- [l-2 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - keepalive: connection org.apache.calcite.avatica.AvaticaJdbc41Factory$AvaticaJdbc41Connection@b89cbf9 is alive
```

At some point I also thought there was an issue with the [connection's stickiness](https://druid.apache.org/docs/latest/querying/sql-jdbc/#connection-stickiness) but the druid i'm using only has a single broker.

**Comment by BT-VT** [comment]
I was seeing the same exception while using a HikariCP pool to query Druid. The fix that worked for me was decreasing the [maxLifetime](https://github.com/brettwooldridge/HikariCP#frequently-used) set in my HikariConfig. I had it set to 15 min, and I noticed when I increased it the connection still failed at the 15 minute mark. I believe there's some other connection time limit equal to 15 min that is enforced on the druid server-side. When my maxLifetime was equal or greater than the druid limit, my HikariCP instance would continue trying to use the same connection after it's expired on the server side and druid would respond with a NoSuchConnectionException. **Lowering the maxLifetime forces the connection to be replaced in the pool before the server-side connection time limit is hit.**

I originally saw this issue while debugging for long periods of time after the connection had been established. I thought it was related to the long idling of the connection while debugging but then I saw it happen while I was running tests that continuously queried druid for over 15 minutes in which case the connection was never idol.
